File: sdk/python/agent_eval/trace.py
"""Trace/Span 上下文管理"""
import time
import uuid
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from agent_eval.client import AgentEval


class SpanContext:
    """Span 上下文管理器"""

    def __init__(
        self,
        trace: "TraceContext",
        name: str,
        kind: str = "other",
        parent_span_id: Optional[str] = None,
    ):
        self._trace = trace
        self.span_id = str(uuid.uuid4())
        self.name = name
        self.kind = kind
        self.parent_span_id = parent_span_id
        self._start_time = time.time()
        self._start_dt = datetime.now(timezone.utc)
        self._input: Optional[str] = None
        self._output: Optional[str] = None
        self._status = "ok"
        self._status_message: Optional[str] = None
        self._attributes: Dict[str, Any] = {}
        self._events: List[Dict[str, Any]] = []
        # LLM
        self._model: Optional[str] = None
        self._prompt_tokens: Optional[int] = None
        self._completion_tokens: Optional[int] = None
        self._temperature: Optional[float] = None
        self._cost: Optional[float] = None
        # Tool
        self._tool_name: Optional[str] = None
        self._tool_input: Optional[str] = None
        self._tool_output: Optional[str] = None
        self._tool_status: Optional[str] = None

    def set_input(self, value: Any) -> "SpanContext":
        self._input = str(value) if value is not None else None
        return self

    def set_output(self, value: Any) -> "SpanContext":
        self._output = str(value) if value is not None else None
        return self

    def set_model(self, model: str) -> "SpanContext":
        self._model = model
        return self

    def set_token_usage(self, prompt: int = 0, completion: int = 0) -> "SpanContext":
        self._prompt_tokens = prompt
        self._completion_tokens = completion
        return self

    def set_temperature(self, temp: float) -> "SpanContext":
        self._temperature = temp
        return self

    def set_cost(self, cost: float) -> "SpanContext":
        self._cost = cost
        return self

    def set_tool(self, name: str, input_data: Any = None, output_data: Any = None, status: str = "success") -> "SpanContext":
        self._tool_name = name
        self._tool_input = str(input_data) if input_data is not None else None
        self._tool_output = str(output_data) if output_data is not None else None
        self._tool_status = status
        return self

    def set_status(self, status: str, message: Optional[str] = None) -> "SpanContext":
        self._status = status
        self._status_message = message
        return self

    def set_attribute(self, key: str, value: Any) -> "SpanContext":
        self._attributes[key] = value
        return self

    def add_event(self, name: str, attributes: Optional[Dict[str, Any]] = None) -> "SpanContext":
        self._events.append({
            "name": name,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "attributes": attributes or {},
        })
        return self

    def __enter__(self) -> "SpanContext":
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        if exc_type:
            self._status = "error"
            self._status_message = str(exc_val)
        self._trace._add_span(self._to_dict())

    def _to_dict(self) -> Dict[str, Any]:
        end_time = time.time()
        latency_ms = int((end_time - self._start_time) * 1000)
        data: Dict[str, Any] = {
            "span_id": self.span_id,
            "trace_id": self._trace.trace_id,
            "name": self.name,
            "kind": self.kind,
            "status": self._status,
            "start_time": self._start_dt.isoformat(),
            "end_time": datetime.now(timezone.utc).isoformat(),
            "latency_ms": latency_ms,
        }
        if self.parent_span_id:
            data["parent_span_id"] = self.parent_span_id
        if self._input is not None:
            data["input"] = self._input
        if self._output is not None:
            data["output"] = self._output
        if self._status_message:
            data["status_message"] = self._status_message
        if self._attributes:
            data["attributes"] = self._attributes
        if self._events:
            data["events"] = self._events
        # LLM
        if self._model:
            data["llm_model"] = self._model
        if self._prompt_tokens is not None:
            data["llm_prompt_tokens"] = self._prompt_tokens
        if self._completion_tokens is not None:
            data["llm_completion_tokens"] = self._completion_tokens
        if self._temperature is not None:
            data["llm_temperature"] = self._temperature
        if self._cost is not None:
            data["llm_cost"] = self._cost
        # Tool
        if self._tool_name:
            data["tool_name"] = self._tool_name
        if self._tool_input is not None:
            data["tool_input"] = self._tool_input
        if self._tool_output is not None:
            data["tool_output"] = self._tool_output
        if self._tool_status:
            data["tool_status"] = self._tool_status
        return data


class TraceContext:
    """Trace 上下文管理器"""

    def __init__(
        self,
        client: "AgentEval",
        name: str,
        session_id: Optional[str] = None,
        user_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
        tags: Optional[List[str]] = None,
    ):
        self._client = client
        self.trace_id = str(uuid.uuid4())
        self.name = name
        self._session_id = session_id
        self._user_id = user_id
        self._metadata = metadata or {}
        self._tags = tags or []
        self._start_time = time.time()
        self._start_dt = datetime.now(timezone.utc)
        self._input: Optional[str] = None
        self._output: Optional[str] = None
        self._status = "ok"
        self._spans: List[Dict[str, Any]] = []
        self._scores: List[Dict[str, Any]] = []

    def set_input(self, value: Any) -> "TraceContext":
        self._input = str(value) if value is not None else None
        return self

    def set_output(self, value: Any) -> "TraceContext":
        self._output = str(value) if value is not None else None
        return self

    def set_status(self, status: str) -> "TraceContext":
        self._status = status
        return self

    def set_metadata(self, key: str, value: Any) -> "TraceContext":
        self._metadata[key] = value
        return self

    def add_tag(self, tag: str) -> "TraceContext":
        self._tags.append(tag)
        return self

    def span(
        self,
        name: str,
        kind: str = "other",
        parent_span_id: Optional[str] = None,
    ) -> SpanContext:
        """创建子 Span"""
        return SpanContext(
            trace=self,
            name=name,
            kind=kind,
            parent_span_id=parent_span_id,
        )

    def score(
        self,
        name: str,
        value: Optional[float] = None,
        string_value: Optional[str] = None,
        comment: Optional[str] = None,
    ) -> "TraceContext":
        """上报 Score"""
        self._scores.append({
            "trace_id": self.trace_id,
            "name": name,
            "value": value,
            "string_value": string_value,
            "comment": comment,
            "source": "sdk",
        })
        return self

    def _add_span(self, span_data: Dict[str, Any]) -> None:
        self._spans.append(span_data)

    def __enter__(self) -> "TraceContext":
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        if exc_type:
            self._status = "error"
        self._flush()

    def end(self) -> None:
        """手动结束 Trace（不使用 context manager 时调用）"""
        self._flush()

    def _flush(self) -> None:
        """将 Trace + Spans + Scores 发送到平台"""
        try:
            end_time = time.time()
            latency_ms = int((end_time - self._start_time) * 1000)

            # 构建 Trace 事件
            trace_event = {
                "type": "trace",
                "trace_id": self.trace_id,
                "agent_id": self._client.agent_id,
                "name": self.name,
                "source": "sdk",
                "status": self._status,
                "start_time": self._start_dt.isoformat(),
                "end_time": datetime.now(timezone.utc).isoformat(),
                "total_latency_ms": latency_ms,
                "session_id": self._session_id,
                "user_id": self._user_id,
                "metadata": self._metadata,
                "tags": self._tags,
            }
            if self._input is not None:
                trace_event["input"] = self._input
            if self._output is not None:
                trace_event["output"] = self._output

            # 聚合统计
            llm_count = sum(1 for s in self._spans if s.get("kind") == "llm")
            tool_count = sum(1 for s in self._spans if s.get("kind") == "tool")
            total_tokens = sum(
                (s.get("llm_prompt_tokens") or 0) + (s.get("llm_completion_tokens") or 0)
                for s in self._spans
            )
            trace_event["llm_call_count"] = llm_count
            trace_event["tool_call_count"] = tool_count
            trace_event["total_tokens"] = total_tokens

            # 发送 Trace
            self._client._transport.enqueue(trace_event)

            # 发送 Spans
            for span_data in self._spans:
                span_event = {"type": "span", **span_data}
                self._client._transport.enqueue(span_event)

            # 发送 Scores
            for score_data in self._scores:
                self._client._transport.send_score(score_data)

        except Exception:
            pass

File: sdk/typescript/src/trace.ts
import { randomUUID } from 'crypto';
import type {
  SpanOptions,
  SpanKind,
  TokenUsage,
  SpanEvent,
  TraceEvent,
  SpanEventData,
  ScoreOptions,
  ToolInfo,
} from './types';
import type { AgentEval } from './client';

/**
 * SpanContext — 表示 Trace 中的一个操作步骤。
 * 支持链式调用和 try-finally 模式。
 */
export class SpanContext {
  readonly spanId: string;
  private trace: TraceContext;
  private name: string;
  private kind: SpanKind;
  private parentSpanId?: string;
  private startTime: Date;
  private endTime?: Date;
  private status: string = 'ok';
  private statusMessage?: string;
  private input?: string;
  private output?: string;
  private attributes: Record<string, unknown> = {};
  private events: SpanEvent[] = [];
  // LLM
  private llmModel?: string;
  private llmPromptTokens?: number;
  private llmCompletionTokens?: number;
  private llmTemperature?: number;
  private llmCost?: number;
  // Tool
  private toolName?: string;
  private toolInput?: string;
  private toolOutput?: string;
  private toolStatus?: string;

  constructor(trace: TraceContext, options: SpanOptions) {
    this.spanId = randomUUID();
    this.trace = trace;
    this.name = options.name;
    this.kind = options.kind || 'other';
    this.parentSpanId = options.parentSpanId;
    this.startTime = new Date();
  }

  setInput(value: unknown): this {
    this.input = typeof value === 'string' ? value : JSON.stringify(value);
    return this;
  }

  setOutput(value: unknown): this {
    this.output = typeof value === 'string' ? value : JSON.stringify(value);
    if (this.output && this.output.length > 5000) {
      this.output = this.output.substring(0, 5000) + '...[truncated]';
    }
    return this;
  }

  setModel(model: string): this {
    this.llmModel = model;
    return this;
  }

  setTokenUsage(usage: TokenUsage): this {
    this.llmPromptTokens = usage.promptTokens;
    this.llmCompletionTokens = usage.completionTokens;
    return this;
  }

  setTemperature(temp: number): this {
    this.llmTemperature = temp;
    return this;
  }

  setCost(cost: number): this {
    this.llmCost = cost;
    return this;
  }

  setTool(tool: ToolInfo): this {
    this.toolName = tool.name;
    this.toolInput = tool.input;
    this.toolOutput = tool.output;
    this.toolStatus = tool.status || 'success';
    return this;
  }

  setStatus(status: string, message?: string): this {
    this.status = status;
    this.statusMessage = message;
    return this;
  }

  setAttribute(key: string, value: unknown): this {
    this.attributes[key] = value;
    return this;
  }

  addEvent(name: string, attributes?: Record<string, unknown>): this {
    this.events.push({
      name,
      timestamp: new Date().toISOString(),
      attributes,
    });
    return this;
  }

  /** 结束 Span 并注册到 Trace */
  end(): void {
    this.endTime = new Date();
    this.trace['_addSpan'](this);
  }

  /** 在回调中执行操作，自动管理 Span 生命周期 */
  async wrap<T>(fn: (span: SpanContext) => Promise<T>): Promise<T> {
    try {
      const result = await fn(this);
      this.end();
      return result;
    } catch (err) {
      this.setStatus('error', err instanceof Error ? err.message : String(err));
      this.end();
      throw err;
    }
  }

  /** 同步版本的 wrap */
  wrapSync<T>(fn: (span: SpanContext) => T): T {
    try {
      const result = fn(this);
      this.end();
      return result;
    } catch (err) {
      this.setStatus('error', err instanceof Error ? err.message : String(err));
      this.end();
      throw err;
    }
  }

  /** 序列化为传输格式 */
  _toEventData(): SpanEventData {
    const latencyMs = this.endTime
      ? this.endTime.getTime() - this.startTime.getTime()
      : undefined;

    const data: SpanEventData['data'] = {
      span_id: this.spanId,
      trace_id: this.trace.traceId,
      name: this.name,
      kind: this.kind,
      status: this.status,
      start_time: this.startTime.toISOString(),
      end_time: this.endTime?.toISOString(),
      latency_ms: latencyMs,
      input: this.input,
      output: this.output,
      parent_span_id: this.parentSpanId,
      attributes: Object.keys(this.attributes).length > 0 ? this.attributes : undefined,
      events: this.events.length > 0 ? this.events : undefined,
      status_message: this.statusMessage,
    };

    // LLM 字段
    if (this.llmModel) data.llm_model = this.llmModel;
    if (this.llmPromptTokens != null) data.llm_prompt_tokens = this.llmPromptTokens;
    if (this.llmCompletionTokens != null) data.llm_completion_tokens = this.llmCompletionTokens;
    if (this.llmTemperature != null) data.llm_temperature = this.llmTemperature;
    if (this.llmCost != null) data.llm_cost = this.llmCost;

    // Tool 字段
    if (this.toolName) data.tool_name = this.toolName;
    if (this.toolInput) data.tool_input = this.toolInput;
    if (this.toolOutput) data.tool_output = this.toolOutput;
    if (this.toolStatus) data.tool_status = this.toolStatus;

    return { type: 'span', data };
  }
}


/**
 * TraceContext — 表示一次完整的追踪。
 * 支持链式调用和 try-finally 模式。
 */
export class TraceContext {
  readonly traceId: string;
  private client: AgentEval;
  private name: string;
  private sessionId?: string;
  private userId?: string;
  private metadata: Record<string, unknown>;
  private tags: string[];
  private startTime: Date;
  private endTime?: Date;
  private status: string = 'ok';
  private input?: string;
  private output?: string;
  private spans: SpanContext[] = [];
  private scores: Array<Omit<ScoreOptions, 'traceId'>> = [];

  constructor(client: AgentEval, options: {
    name: string;
    sessionId?: string;
    userId?: string;
    metadata?: Record<string, unknown>;
    tags?: string[];
  }) {
    this.traceId = randomUUID();
    this.client = client;
    this.name = options.name;
    this.sessionId = options.sessionId;
    this.userId = options.userId;
    this.metadata = options.metadata || {};
    this.tags = options.tags || [];
    this.startTime = new Date();
  }

  setInput(value: unknown): this {
    this.input = typeof value === 'string' ? value : JSON.stringify(value);
    return this;
  }

  setOutput(value: unknown): this {
    this.output = typeof value === 'string' ? value : JSON.stringify(value);
    if (this.output && this.output.length > 5000) {
      this.output = this.output.substring(0, 5000) + '...[truncated]';
    }
    return this;
  }

  setStatus(status: string): this {
    this.status = status;
    return this;
  }

  setMetadata(key: string, value: unknown): this {
    this.metadata[key] = value;
    return this;
  }

  addTag(tag: string): this {
    if (!this.tags.includes(tag)) {
      this.tags.push(tag);
    }
    return this;
  }

  /** 创建子 Span */
  span(options: SpanOptions | string): SpanContext {
    const opts = typeof options === 'string' ? { name: options } : options;
    return new SpanContext(this, opts);
  }

  /** 在 Trace 上下文中记录 Score（Trace 结束时批量发送） */
  score(name: string, value?: number, options?: { stringValue?: string; comment?: string }): this {
    this.scores.push({
      name,
      value,
      stringValue: options?.stringValue,
      comment: options?.comment,
    });
    return this;
  }

  /** 结束 Trace 并发送所有数据 */
  async end(): Promise<void> {
    this.endTime = new Date();
    await this._flush();
  }

  /** 在回调中执行操作，自动管理 Trace 生命周期 */
  async wrap<T>(fn: (trace: TraceContext) => Promise<T>): Promise<T> {
    try {
      const result = await fn(this);
      await this.end();
      return result;
    } catch (err) {
      this.setStatus('error');
      await this.end();
      throw err;
    }
  }

  /** 内部方法：添加已结束的 Span */
  private _addSpan(span: SpanContext): void {
    this.spans.push(span);
  }

  /** 内部方法：Flush 所有数据到传输层 */
  private async _flush(): Promise<void> {
    const transport = this.client['transport'];
    if (!transport) return;

    // 统计聚合
    let llmCallCount = 0;
    let toolCallCount = 0;
    let totalTokens = 0;

    for (const span of this.spans) {
      const eventData = span._toEventData();
      const d = eventData.data;
      if (d.kind === 'llm') llmCallCount++;
      if (d.kind === 'tool') toolCallCount++;
      if (d.llm_prompt_tokens) totalTokens += d.llm_prompt_tokens;
      if (d.llm_completion_tokens) totalTokens += d.llm_completion_tokens;
    }

    const latencyMs = this.endTime
      ? this.endTime.getTime() - this.startTime.getTime()
      : undefined;

    // 发送 Trace 事件
    const traceEvent: TraceEvent = {
      type: 'trace',
      data: {
        trace_id: this.traceId,
        agent_id: this.client['agentId'],
        name: this.name,
        status: this.status,
        start_time: this.startTime.toISOString(),
        end_time: this.endTime?.toISOString(),
        latency_ms: latencyMs,
        input: this.input,
        output: this.output,
        session_id: this.sessionId,
        user_id: this.userId,
        metadata: Object.keys(this.metadata).length > 0 ? this.metadata : undefined,
        tags: this.tags.length > 0 ? this.tags : undefined,
        source: 'sdk',
        llm_call_count: llmCallCount,
        tool_call_count: toolCallCount,
        total_tokens: totalTokens,
      },
    };

    transport.enqueue(traceEvent);

    // 发送 Span 事件
    for (const span of this.spans) {
      transport.enqueue(span._toEventData());
    }

    // 发送 Score
    for (const s of this.scores) {
      await transport.sendScore({
        traceId: this.traceId,
        name: s.name,
        value: s.value,
        stringValue: s.stringValue,
        comment: s.comment,
        source: 'sdk',
      });
    }
  }
}
